generator/camera_info_collector.cpp

```diff
@@ -0,0 +1,340 @@
+  uint32_t segmentIdOfBestFeatureId;
+  auto minDist = std::numeric_limits<double>::max();
+  double coef = 0;
+
+  // Look at each segment of roads and find the closest.
+  auto const updClosestFeatureCallback = [&](FeatureType & ft) {
+    if (ft.GetFeatureType() != feature::GEOM_LINE)
+      return;
+
+    if (!routing::IsCarRoad(feature::TypesHolder(ft)))
+      return;
+
+    auto curMinDist = std::numeric_limits<double>::max();
+
+    ft.ParseGeometry(scales::GetUpperScale());
+    std::vector<m2::PointD> points(ft.GetPointsCount());
+    for (size_t i = 0; i < points.size(); ++i)
+      points[i] = ft.GetPoint(i);
+
+    routing::FollowedPolyline polyline(points.begin(), points.end());
+    m2::RectD const rect =
+        MercatorBounds::RectByCenterXYAndSizeInMeters(m_center, kSearchCameraRadiusMeters);
+    auto bestSegment = polyline.UpdateProjection(rect);
+    double tmpCoef = 0;
+
+    if (bestSegment.IsValid())
+    {
+      static double constexpr kEps = 1e-6;
+      auto const p1 = polyline.GetPolyline().GetPoint(bestSegment.m_ind);
+      auto const p2 = polyline.GetPolyline().GetPoint(bestSegment.m_ind + 1);
+
+      if (AlmostEqualAbs(p1, p2, kEps))
+        return;
+
+      m2::ParametrizedSegment<m2::PointD> st(p1, p2);
+      auto const cameraProjOnSegment = st.ClosestPointTo(m_center);
+      curMinDist = MercatorBounds::DistanceOnEarth(cameraProjOnSegment, m_center);
+
+      tmpCoef = MercatorBounds::DistanceOnEarth(p1, cameraProjOnSegment) /
+                MercatorBounds::DistanceOnEarth(p1, p2);
+    }
+
+    if (curMinDist < minDist && curMinDist < kMaxDistFromCameraToClosestSegmentMeters)
+    {
+      minDist = curMinDist;
+      bestFeatureId = ft.GetID().m_index;
+      segmentIdOfBestFeatureId = static_cast<uint32_t>(bestSegment.m_ind);
+      coef = tmpCoef;
+    }
+  };
+
+  dataSource.ForEachInRect(
+      updClosestFeatureCallback,
+      MercatorBounds::RectByCenterXYAndSizeInMeters(m_center, kSearchCameraRadiusMeters),
+      scales::GetUpperScale());
+
+  if (minDist != std::numeric_limits<double>::max())
+    m_ways.emplace_back(bestFeatureId, segmentIdOfBestFeatureId, coef);
+}
+
+boost::optional<std::pair<double, uint32_t>> CamerasInfoCollector::Camera::FindMyself(
+    uint32_t wayId, FrozenDataSource const & dataSource, MwmSet::MwmId const & mwmId) const
+{
+  double coef = 0;
+  bool isRoad = true;
+  uint32_t result = 0;
+
+  auto const readFeature = [&coef, &result, this, &isRoad, wayId](FeatureType & ft) {
+    bool found = false;
+    isRoad = routing::IsRoad(feature::TypesHolder(ft));
+    if (!isRoad)
+      return;
+
+    auto const forEachPoint = [&result, &found, this](m2::PointD const & pt) {
+      if (found)
+        return;
+
+      if (AlmostEqualAbs(m_center, pt, kEqualityEps))
+        found = true;
+      else
+        ++result;
+    };
+
+    ft.ForEachPoint(forEachPoint, scales::GetUpperScale());
+
+    CHECK(found, ("Cannot find camera point at the feature:", wayId));
```

Comment:
Doesn't wayId correspond to Way Id? Or is it really a feature's id/number?

Reply:
Is it osm way's id or mapsme's feature id?

Reply:
mapsme feature id